server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java

```diff
@@ -193,23 +190,32 @@ private List<StoreFileMetadata> getFiles(CheckpointInfoResponse checkpointInfo)
         return diff.missing;
     }
 
-    private void finalizeReplication(CheckpointInfoResponse checkpointInfoResponse) throws OpenSearchCorruptionException {
-        // TODO: Refactor the logic so that finalize doesn't have to be invoked for remote store as source
-        if (source instanceof RemoteStoreReplicationSource) {
-            state.setStage(SegmentReplicationState.Stage.FINALIZE_REPLICATION);
-            return;
-        }
+    private void finalizeReplication(CheckpointInfoResponse checkpointInfoResponse, GetSegmentFilesResponse getSegmentFilesResponse)
+        throws OpenSearchCorruptionException {
         cancellableThreads.checkForCancel();
         state.setStage(SegmentReplicationState.Stage.FINALIZE_REPLICATION);
         Store store = null;
         try {
             store = store();
             store.incRef();
+            Map<String, String> tempFileNames = null;
+            if (this.indexShard.indexSettings().isRemoteStoreEnabled() == true) {
```

Comment:
I know why you have to do this here, but thinking we can refactor this further to remove these divergent paths in the target.

Can we move the implementation specific logic under the ReplicationSource interface?

Perhaps we can reuse `getSegmentFiles` here without requiring a new function and pass this data back to the target in the GetSegmentFilesResponse or change it to return a different response object?  This would require some more refactoring to get the MultiFileWriter etc into the PrimaryShardReplicationSource.  We can take this as a follow up?

Reply:
Yes, I think we can take this up in follow up. 

I spent some time into looking through the changes needed for this. There is one core concern with this refactor is that moving `MultiFileWriter` to `ReplicationSoruce` (PrimaryShardReplicationSource) seems not logically right as it logically belongs to `ReplicationTarget`. We re-used the ReplicationTarget from recovery side of things, it appears now there is need to change that. This change needs more discussion and thus better suited in a follow up. Will keep the comment unresolve.